### PR TITLE
Model Changes (added new and resolved bug) 

### PR DESCRIPTION
**Pull request by victorlazio109**
new model added (unet + context) 
sampling function added
resolved model loading bug on > 1 GPU device

### CODE REVIEW

File: models/model_choice.py [hunk 1]
```diff
@@ -20,8 +23,9 @@ def load_checkpoint(filename):
     '''
     try:
         print(f"=> loading model '{filename}'\n")
-        checkpoint = torch.load(filename) if torch.cuda.is_available() else torch.load(filename, map_location='cpu')
+        # checkpoint = torch.load(filename) if torch.cuda.is_available() else torch.load(filename, map_location='cpu')
```
Comment (CharlesAuthier) [inline models/model_choice.py:26]: Maybe delete this line since you put it after

File: models/unet_context.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        ctx4_l = self.ctx4_l(ctx3_l)
+        ctx5_l = self.ctx5_l(ctx4_l)
+        ctx7_l = self.ctx7_l(ctx5_l)
+        conv_l = self.conv_l(ctx7_l)
+
+        return conv_l
+
+
+class ContextualUnet(nn.Module):
+
```
Comment (CharlesAuthier) [inline models/unet_context.py:64]: Maybe add a little doc plus link for the article (if it have one)

File: train_segmentation.py [hunk 3]
```diff
@@ -107,13 +104,17 @@ def create_dataloader(samples_folder, batch_size, num_devices, params):
     num_workers = num_devices * 4 if num_devices > 1 else 4
 
     # Shuffle must be set to True.
-    trn_dataloader = DataLoader(trn_dataset, batch_size=batch_size, num_workers=num_workers, shuffle=True,
+    samples_weight = torch.from_numpy(samples_weight)
+    sampler = torch.utils.data.sampler.WeightedRandomSampler(samples_weight.type('torch.DoubleTensor'),
+                                                             len(samples_weight))
+
+    trn_dataloader = DataLoader(trn_dataset, batch_size=batch_size, num_workers=num_workers, sampler=sampler,
```
Comment (CharlesAuthier) [inline train_segmentation.py:111]: no shuffle?

File: train_segmentation.py [hunk 4]
```diff
@@ -140,8 +141,15 @@ def get_num_samples(samples_path, params):
         else:
             with h5py.File(samples_path.joinpath(f"{i}_samples.hdf5"), "r") as hdf5_file:
                 num_samples[i] = len(hdf5_file['map_img'])
+                if i == 'trn':
```
Comment (CharlesAuthier) [inline train_segmentation.py:144]: Is that the more efficient way to get the num of sample? the for loop seem to be to much and heavy to compute.

File: train_segmentation.py (hunk 3 above)
Comment (victorlazio109) [reply train_segmentation.py:111]: With the sampler, you can't use shuffle. The samples are drawn based on the weights computed in random order.

File: train_segmentation.py (hunk 4 above)
Comment (victorlazio109) [reply train_segmentation.py:144]: I haven't thought much about it. any suggestions?

Comment (CharlesAuthier) [reply train_segmentation.py:144]: give me the afternoon to check and I come back to you

File: models/model_choice.py [hunk 5]
```diff
@@ -32,6 +35,56 @@ def load_checkpoint(filename):
...
+    if num_classes == 1 and len(weights) == 2:
+        warnings.warn("got two class weights for single class defined in configuration file; will assume index 0 = background")
+    elif num_classes != len(weights):
+        raise ValueError('The number of class weights in the configuration file is different than the number of classes')
+
+
+def set_hyperparameters(params, num_classes, model, checkpoint, dontcare_val):
+    """
+    Function to set hyperparameters based on values provided in yaml config file.
+    Will also set model to GPU, if available.
```
Comment (remtav) [inline models/model_choice.py:53]: I don't see any part of this function dedicated to setting model to GPU, am I blind? If not, we should update the docstring. This statement may date from a while ago.

File: models/model_choice.py [hunk 6]
```diff
@@ -141,6 +201,8 @@ def net(net_params, num_channels, inference=False):
             in_channels=num_bands,
             classes=num_channels,
             activation=None)
+    elif model_name == 'unet_context':
+        model = unet_context.ContextualUnet(num_bands, num_channels)
```
Comment (remtav) [inline models/model_choice.py:205]: nice models collection. Gotta catch 'em all!

File: models/model_choice.py [hunk 7]
```diff
@@ -153,19 +215,53 @@ def net(net_params, num_channels, inference=False):
...
+            checkpoint = None
+        assert num_devices is not None and num_devices >= 0, "missing mandatory num gpus parameter"
+        # list of GPU devices that are available and unused. If no GPUs, returns empty list
+        lst_device_ids = get_device_ids(num_devices) if torch.cuda.is_available() else []
+        num_devices = len(lst_device_ids) if lst_device_ids else 0
+        device = torch.device(f'cuda:{lst_device_ids[0]}' if torch.cuda.is_available() and lst_device_ids else 'cpu')
+        print(f"Number of cuda devices requested: {net_params['global']['num_gpus']}. Cuda devices available: {lst_device_ids}\n")
+        if num_devices == 1:
+            print(f"Using Cuda device {lst_device_ids[0]}\n")
+        elif num_devices > 1:
```
Comment (remtav) [inline models/model_choice.py:243]: I understand. This is where model is pushed to GPU if available. Conforms docstring in load_checkpoint should be updated.

File: train_segmentation.py (hunk 3 above)
Comment (remtav) [reply train_segmentation.py:111]: This sampler is really a nice addition. Thanks.

File: train_segmentation.py [hunk 8]
```diff
@@ -107,13 +104,17 @@ def create_dataloader(samples_folder, batch_size, num_devices, params):
...
+    samples_weight = torch.from_numpy(samples_weight)
+    sampler = torch.utils.data.sampler.WeightedRandomSampler(samples_weight.type('torch.DoubleTensor'),
+                                                             len(samples_weight))
+
+    trn_dataloader = DataLoader(trn_dataset, batch_size=batch_size, num_workers=num_workers, sampler=sampler,
                                 drop_last=True)
     # Using batch_metrics with shuffle=False on val dataset will always mesure metrics on the same portion of the val samples.
     # Shuffle should be set to True.
-    val_dataloader = DataLoader(val_dataset, batch_size=batch_size, num_workers=num_workers, shuffle=True,
+    val_dataloader = DataLoader(val_dataset, batch_size=1, num_workers=num_workers, shuffle=True,
```
Comment (remtav) [inline train_segmentation.py:115]: Why is batch_size being reduced to 1?
Also, I recently created issue #168 . Do you want to address it in this PR? I can take care of it in my next PR otherwise.

File: train_segmentation.py [hunk 9]
```diff
@@ -487,59 +504,6 @@ def main(params, config_path):
...
-        device = torch.device(f'cuda:0' if torch.cuda.is_available() and lst_device_ids else 'cpu')
-        model.to(device)
-    model, criterion, optimizer, lr_scheduler = set_hyperparameters(params,
-                                                                    num_classes_corrected,
-                                                                    model,
-                                                                    checkpoint,
-                                                                    dontcare)
-
-    criterion = criterion.to(device)
-
```
Comment (remtav) [inline train_segmentation.py:542]: A lot of code moving around. Have you tested extensively to make sure nothing might break? Is there a general reason for these movements? Getting new features to work, structuring things more logically, etc.?

File: models/model_choice.py (hunk 7 above)
Comment (victorlazio109) [reply models/model_choice.py:248]: Done

File: train_segmentation.py [hunk 10]
```diff
@@ -112,7 +112,7 @@ def create_dataloader(samples_folder, batch_size, num_devices, params):
                                 drop_last=True)
     # Using batch_metrics with shuffle=False on val dataset will always mesure metrics on the same portion of the val samples.
```
Comment (mpelchat04) [inline train_segmentation.py:113]: a) adjust the two following commented lines to reflect why it should be set to shuffle=False  
or  
b) delete those two commented lines.

**Review by mpelchat04 — APPROVED** [review]

**Review by bstdenis — DISMISSED** [review]

File: models/model_choice.py [hunk 11]
```diff
@@ -68,6 +71,55 @@ def load_checkpoint(filename):
...
+    :return: model, criterion, optimizer, lr_scheduler, num_gpus
+    """
+    # set mandatory hyperparameters values with those in config file if they exist
+    lr = get_key_def('learning_rate', params['training'], None, "missing mandatory learning rate parameter")
+    weight_decay = get_key_def('weight_decay', params['training'], None, "missing mandatory weight decay parameter")
+    step_size = get_key_def('step_size', params['training'], None, "missing mandatory step size parameter")
+    gamma = get_key_def('gamma', params['training'], None, "missing mandatory gamma parameter")
+
+    # optional hyperparameters. Set to None if not in config file
+    class_weights = torch.tensor(params['training']['class_weights']) if params['training']['class_weights'] else None
```
Comment (remtav) [inline models/model_choice.py:103]: the get_key_def function can also deal with optional hyperparameters, even [when parameter is not even found](https://github.com/NRCan/geo-deep-learning/blob/6852729d0ec03c0c2a6465b7a68039181c8bcbda/utils/utils.py#L150) in the corresponding yaml section. Why not use get_key_def here also?

File: models/model_choice.py [hunk 12]
```diff
@@ -68,6 +71,55 @@ def load_checkpoint(filename):
...
+        num_classes: number of classes defined in the configuration file
+        weights: weights defined in the configuration file
+    """
+    if num_classes == 1 and len(weights) == 2:
+        warnings.warn("got two class weights for single class defined in configuration file; will assume index 0 = background")
+    elif num_classes != len(weights):
+        raise ValueError('The number of class weights in the configuration file is different than the number of classes')
+
+
+def set_hyperparameters(params, num_classes, model, checkpoint, dontcare_val):
```
Comment (remtav) [inline models/model_choice.py:86]: We have a `set_hyperparameters()` function [in train_segmentation.py](https://github.com/NRCan/geo-deep-learning/blob/6852729d0ec03c0c2a6465b7a68039181c8bcbda/train_segmentation.py#L147). Though it shouldn't be a problem for autocompletion, for example, in our IDEs, should we consider a different name? ex.: `set_model_hyperparam()`

File: models/model_choice.py [hunk 13]
```diff
@@ -68,6 +71,55 @@ def load_checkpoint(filename):
...
+    """
+    # set mandatory hyperparameters values with those in config file if they exist
+    lr = get_key_def('learning_rate', params['training'], None, "missing mandatory learning rate parameter")
+    weight_decay = get_key_def('weight_decay', params['training'], None, "missing mandatory weight decay parameter")
+    step_size = get_key_def('step_size', params['training'], None, "missing mandatory step size parameter")
+    gamma = get_key_def('gamma', params['training'], None, "missing mandatory gamma parameter")
+
+    # optional hyperparameters. Set to None if not in config file
+    class_weights = torch.tensor(params['training']['class_weights']) if params['training']['class_weights'] else None
+    if params['training']['class_weights']:
```
Comment (remtav) [inline models/model_choice.py:104]: I'd suggest, once again, using get_key_def() since it's meant to be a generic function for reading parameters from yaml. Same applies for yaml parameters stored in variables below (loss, optimizer, num_gpus, etc.).

One neat feature you could use here, for example, is the [`expected_type` argument](https://github.com/NRCan/geo-deep-learning/blob/6852729d0ec03c0c2a6465b7a68039181c8bcbda/utils/utils.py#L155) of `get_key_def()`, i.e. expect a `typing.List` (right?) here. It would also help with the exception handling since `verify_weights()` doesn't validate the kind of object being passed in the `weights` argument.

I just saw that I didn't document the expected type argument it in the docstrings. Bad karma.

File: train_segmentation.py (hunk 9 above)
Comment (remtav) [reply train_segmentation.py:542]: hey! I had the same though again :) any answer @victorlazio109 ?

**Review by CharlesAuthier — DISMISSED** [review]

File: train_segmentation.py (hunk 9 above)
Comment (remtav) [reply train_segmentation.py:542]: Oh I see! I hadn't seen your previous answer. Thanks.

File: models/model_choice.py (hunk 11 above)
Comment (remtav) [reply models/model_choice.py:103]: No no go ahead. These remarks are minor and indeed, don't even cover code that you changed. I was coded this way (withouth get_key_def to begin with).

**Review by remtav — APPROVED** [review]

**Review by mpelchat04 — APPROVED** [review]

**Review by bstdenis — APPROVED** [review]

**Review by CharlesAuthier — APPROVED** [review]
